### PR TITLE
[AIRFLOW-4292] Cleanup and improve SLA code

### PR DESCRIPTION
**Pull request by serkef**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-4292
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [ ] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
No functionality change - current unit tests should suffice.

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

File: airflow/jobs.py [hunk 1]
```diff
@@ -660,86 +660,88 @@ def manage_slas(self, dag, session=None):
...
-            )
-            blocking_tis = []
-            for ti in qry:
-                if ti.task_id in dag.task_ids:
-                    ti.task = dag.get_task(ti.task_id)
-                    blocking_tis.append(ti)
-                else:
-                    session.delete(ti)
-                    session.commit()
+        if not slas:
```
Comment (andriisoldatenko) [inline airflow/jobs.py:673]: what are the purpose of this change?

**Comment by andriisoldatenko** [comment]
@serkef can you please make sure you have pass all checks?

File: airflow/jobs.py (hunk 1 above)
Comment (serkef) [reply airflow/jobs.py:673]: Since there is no else, we can reverse the check to reduce the indentation and make it more readable.

**Comment by serkef** [comment]
> @serkef can you please make sure you have pass all checks?

@andriisoldatenko Cleaned by remerging. Maybe there was a problematic test that got removed?

File: airflow/jobs.py [hunk 2]
```diff
@@ -622,6 +622,7 @@ def manage_slas(self, dag, session=None):
             self.log.info("Skipping SLA check for %s because no tasks in DAG have SLAs", dag)
             return
 
+        # Find the last executed tasks
```
Comment (PhilipGarnero) [inline airflow/jobs.py:627]: If I'm not mistaken, this is not find the last executed tasks but rather, find the last succesful or skipped task instances.
Which by the way is a bit weird to base SLAs from only successful tasks instances. What if the the whole history has failed ? We don't get any SLA ? I think we need to change the filter down below too so it uses any task that hasn't be scheduled yet as a basis.

Comment (PhilipGarnero) [reply airflow/jobs.py:627]: Nevermind, #3584 adresses the problem perfectly.

Comment (serkef) [reply airflow/jobs.py:627]: Exactly, I mention on the ticket there are couple of open issues regarding this topic and this is my first cleanup attempt, without affecting functionality.

**Comment by serkef** [comment]
Closing, merging is quite hard at this stage. 
@PhilipGarnero, @andriisoldatenko, can you give it another look to #5517, includes the same changes? I will close this one.

**Pull request closed without merging**